pkg/kudoctl/packages/types.go

```diff
@@ -43,9 +64,151 @@ func (p Parameter) Less(x, y int) bool {
+	Error          *string
+}
+
+// TemplateNodes is a map of template files to template nodes
+type TemplateNodes map[string]Nodes
+
+// Nodes converts a set of Templates to TemplateNodes which is a map of file names to template nodes
+func (ts Templates) Nodes() TemplateNodes {
+
+	tNodes := TemplateNodes{}
+
+	for fname, file := range ts {
+		// template nodes to be collected in a set
+		// fresh for each file
+		nodeSet := map[string]bool{}
+		t := template.New(fname)
+		// parse 1 template
+		tplate, err := t.Parse(file)
+		if err != nil {
+			errMsg := fmt.Sprintf("template file %q reports the following error: %v", fname, err)
+			n := Nodes{
+				Error: &errMsg,
+			}
+			tNodes[fname] = n
+
+			continue
+		}
+		// cycle through all the nodes and collect Action nodes
+		nodes := walkNodes(tplate.Root, fname)
+		for _, node := range nodes {
+			tnode := trimNode(node)
+			nodeSet[tnode] = true
+		}
+
+		n := Nodes{
+			Parameters:     parameters(nodeSet),
+			ImplicitParams: implicits(nodeSet),
+		}
+		tNodes[fname] = n
+	}
+
+	return tNodes
+}
+
+//walkNodes walks the nodes of a template providing an array of parameters
+func walkNodes(node parse.Node, fname string) []string {
+	switch node := node.(type) {
+	case *parse.ActionNode:
+		return []string{trimNode(node.String())}
+	//	if and with operate the same however we can't fail through in type switch
+	case *parse.IfNode:
+		list := walkNodes(node.List, fname)
+		elist := walkNodes(node.ElseList, fname)
+		list = append(list, elist...)
+		clist := walkPipes(node.Pipe)
+		return append(list, clist...)
+	case *parse.WithNode:
+		list := walkNodes(node.List, fname)
+		elist := walkNodes(node.ElseList, fname)
+		list = append(list, elist...)
+		clist := walkPipes(node.Pipe)
+		return append(list, clist...)
+	case *parse.ListNode:
+		list := []string{}
+		if node == nil {
+			return list
+		}
+		for _, n := range node.Nodes {
+			list = append(list, walkNodes(n, fname)...)
+		}
+		return list
+	case *parse.RangeNode: // no support for Range, Template or TextNodes
+	case *parse.TemplateNode:
+	case *parse.TextNode:
+	default:
+		clog.V(2).Printf("file %q has unknown node: %s", fname, node)
+	}
+	return []string{}
+}
+
+// walkPipes walks the pipes of specific block types which may contain params
+func walkPipes(node *parse.PipeNode) []string {
+	params := []string{}
+	for _, cmd := range node.Cmds {
+		for _, arg := range cmd.Args {
+			switch arg.(type) {
+			case *parse.FieldNode:
+				if strings.HasPrefix(arg.String(), ".Params") {
+					params = append(params, arg.String())
+				}
+			}
+		}
+	}
+	return params
+}
+
+// parameters takes a map and returns Nodes (which is a []string).
+// map is used for set functionality (a goism)
+// parameters converts array of Nodes to just nodes of parameters (those prefixed with "Params.") and strips the prefix
+func parameters(nodes map[string]bool) []string {
+
+	fields := []string{}
+
+	for k := range nodes {
+		if strings.Contains(k, "Params.") {
```

Comment:
Is there something like "strings.startsWith"? Contains could trigger a false positive, i guess

Reply:
`Prefix` yeah... thanks!